Improve accessibility of fuzzy file finder

olafurpg opened the pull request:
Following up on the suggestion in https://github.com/sourcegraph/sourcegraph/issues/21201#issuecomment-856944021 

This PR missing HTML attributes for accessibility in the fuzzy finder component. I tried to roughly use the same attribute values as the HTML structure of the Algolia search box on this page here https://scalameta.org/metals/

felixfbecker commented:
Wow awesome @olafurpg! Great idea looking at what Algolia does, it looks right from a first look. I can't test with VoiceOver unfortunately because of the issue in Storybook but would be happy to do so if that can be fixed!

felixfbecker commented:
Besides the input, is the modal using an existing Modal component from our codebase that implements accessibility (close on <kbd>esc</kbd>, `role="dialog"`, focus trapping)? That would be the other accessibility concern to make sure.

olafurpg commented:
> I can't test with VoiceOver unfortunately because of the issue in Storybook but would be happy to do so if that can be fixed!

Is that issue related to the fuzzy finder in particular or storybook in general?

> Besides the input, is the modal using an existing Modal component from our codebase that implements accessibility (close on esc, role="dialog", focus trapping)? That would be the other accessibility concern to make sure.

There's an unresolved item in the followup issue to use the `<Dialog` modal instead of the hand-rolled implementation.

olafurpg commented:
nvm, I just saw your comment in the other PR about chromatic. I'll address there.

olafurpg commented:
Will update this PR later once the design has landed into main

felixfbecker commented:
One thing I already noticed: Close button is missing `aria-label` (only has an icon). Might go away if using the reusable component but I'm not sure.

olafurpg commented:
@felixfbecker I replaced the custom modal with `<Dialog>` but kept the close button for fidelity with the Figma design proposal. I added `aria-label="Close"` to the close button.

olafurpg commented:
This PR is ready for a second round of review. I rebased on top of main after the initial design landed.

eseliger commented on client/web/src/components/fuzzyFinder/FuzzyModal.module.scss:
```diff
@@ -1,11 +1,8 @@
-    bottom: 0;
-    background-color: rgba(0, 0, 0, 0.5);
+    padding: 0;
+    border-radius: var(--border-radius);
+    border: 1px solid var(--border-color-2);
+    background-color: transparent;
```
Why is there the need to overwrite the modal default here?

olafurpg replied:
Good question, I'm not sure. The default background was white and this change was needed to remove tiny white corners.

![CleanShot 2021-06-10 at 16 23 03](https://user-images.githubusercontent.com/1408093/121542283-52a49900-ca08-11eb-81ee-c11ce746d369.png)

Any suggestion to avoid this?

eseliger replied:
I'll give it a try locally soon

eseliger replied:
Can you give [this PR](https://github.com/sourcegraph/sourcegraph/pull/21991) a try and merge it in here if that works for you? I've tried to get away from the overwrites and don't see the weird white border locally anymore. I'll be out next week so if you don't get to it before then feel free to merge my PR in here if you like, no need to wait for me.

olafurpg replied:
Thanks! I merged your PR into this branch

olafurpg commented:
Could someone help unblock the pending UI Review? 😇

eseliger commented:
Approved! No idea why the backdrop changed, did you verify it still has one in the app?

olafurpg commented:
I just tested the component locally and the backdrop is looking OK (if I understand correctly)

![image](https://user-images.githubusercontent.com/1408093/121685589-67446800-cac0-11eb-92e0-8af7dc35db13.png)

felixfbecker commented:
Just tested this on dotcom and the selected item was not announced. Checking the PR I don't see `aria-selected` being applied to the selected element, but I could have sworn I saw that in a prior version of the PR? Maybe I just saw it on Algolia search box. Each option also needs `role="option"` to be listable by the screen reader.

The easiest way you can test this yourself if you're on a mac is by starting VoiceOver with <kbd>cmd</kbd>+<kbd>F5</kbd> while the file picker is open. It should already start at the input since it's focused. If the voice is overwhelming, just single-tap <kbd>ctrl</kbd> to stop the current read. Then when using up and down arrows the selected option should get announced. VO also says to get a list of options, press <kbd>ctrl</kbd>+<kbd>option</kbd>+<kbd>space</kbd>, which presumably should announce all options, but currently, it doesn't announce anything.

olafurpg commented:
@felixfbecker good catch! I opened #22054 to fix this issue. The original version of this PR did indeed have `role=option` and `aria-selected=true`, but that diff got lost when I rebased the PR on top of main (I manually re-typed the diff instead of using `git-rebase` because there were so many conflicts and the diff was so small, I still messed it up 😅 ).
